apps/hash-frontend/src/shared/notifications-context.tsx

```diff
@@ -0,0 +1,423 @@
+  getRoots,
+} from "@local/hash-subgraph/stdlib";
+import { extractBaseUrl } from "@local/hash-subgraph/type-system-patch";
+import {
+  createContext,
+  FunctionComponent,
+  PropsWithChildren,
+  useCallback,
+  useContext,
+  useMemo,
+} from "react";
+
+import { useBlockProtocolUpdateEntity } from "../components/hooks/block-protocol-functions/knowledge/use-block-protocol-update-entity";
+import {
+  StructuralQueryEntitiesQuery,
+  StructuralQueryEntitiesQueryVariables,
+} from "../graphql/api-types.gen";
+import { structuralQueryEntitiesQuery } from "../graphql/queries/knowledge/entity.queries";
+import { constructMinimalUser, MinimalUser } from "../lib/user-and-org";
+import { useAuthInfo } from "../pages/shared/auth-info-context";
+
+export type PageMentionNotification = {
+  kind: "page-mention";
+  entity: Entity<MentionNotificationProperties>;
+  occurredInPage: Entity<PageProperties>;
+  occurredInBlock: Entity<BlockProperties>;
+  occurredInText: Entity<TextProperties>;
+  triggeredByUser: MinimalUser;
+};
+
+export type CommentMentionNotification = {
+  kind: "comment-mention";
+  occurredInComment: Entity<CommentProperties>;
+} & Omit<PageMentionNotification, "kind">;
+
+export type NewCommentNotification = {
+  kind: "new-comment";
+  entity: Entity<CommentNotificationProperties>;
+  occurredInPage: Entity<PageProperties>;
+  occurredInBlock: Entity<BlockProperties>;
+  triggeredByComment: Entity<CommentProperties>;
+  triggeredByUser: MinimalUser;
+};
+
+export type CommentReplyNotification = {
+  kind: "comment-reply";
+  repliedToComment: Entity<CommentProperties>;
+} & Omit<NewCommentNotification, "kind">;
+
+export type Notification =
+  | PageMentionNotification
+  | CommentMentionNotification
+  | NewCommentNotification
+  | CommentReplyNotification;
+
+export type NotificationsContextValues = {
+  notifications?: Notification[];
+  loading: boolean;
+  refetch: () => Promise<void>;
+  markNotificationAsRead: (params: {
+    notification: Notification;
+  }) => Promise<void>;
+};
+
+export const NotificationsContext =
+  createContext<null | NotificationsContextValues>(null);
+
+export const useNotifications = () => {
+  const notificationsContext = useContext(NotificationsContext);
+
+  if (!notificationsContext) {
+    throw new Error("Context missing");
+  }
+
+  return notificationsContext;
+};
+
+const isLinkAndRightEntityWithLinkType =
+  (linkEntityTypeId: VersionedUrl) =>
+  ({ linkEntity }: LinkEntityAndRightEntity) =>
+    linkEntity[0] && linkEntity[0].metadata.entityTypeId === linkEntityTypeId;
+
+export const NotificationsContextProvider: FunctionComponent<
+  PropsWithChildren
+> = ({ children }) => {
+  const { authenticatedUser } = useAuthInfo();
+
+  const {
+    data: notificationsData,
+    loading,
+    refetch: refetchQuery,
+  } = useQuery<
+    StructuralQueryEntitiesQuery,
+    StructuralQueryEntitiesQueryVariables
+  >(structuralQueryEntitiesQuery, {
+    variables: {
```

Comment:
Poll for new notifications. 5s seems reasonable
```suggestion
    pollInterval: 5_000,
    variables: {
```